stratosphere/ncm/source/lr_registeredlocationresolver.cpp

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2019 Adubbz
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "lr_registeredlocationresolver.hpp"
+
+namespace ams::lr {
+
+    RegisteredLocationResolverInterface::~RegisteredLocationResolverInterface() {
+        /* Ensure entries are deallocated */
+        this->ClearRedirections();
+    }
+
+    void RegisteredLocationResolverInterface::ClearRedirections(u32 flags) {
+        this->html_docs_redirector.ClearRedirections();
+        this->program_redirector.ClearRedirections();
+    }
+
+    void RegisteredLocationResolverInterface::RegisterPath(const Path& path, impl::RegisteredLocations<ncm::ProgramId, RegisteredLocationResolverInterface::MaxRegisteredLocations>* locations, ncm::ProgramId id, ncm::ProgramId owner_id) {
+        if (!locations->Register(id, path, owner_id)) {
+            locations->Clear();
+            locations->Register(id, path, owner_id);
+        }
+    }
+
+    bool RegisteredLocationResolverInterface::ResolvePath(Path* out, impl::LocationRedirector* redirector, impl::RegisteredLocations<ncm::ProgramId, RegisteredLocationResolverInterface::MaxRegisteredLocations>* locations, ncm::ProgramId id) {
+        if (!redirector->FindRedirection(out, id)) {
+            if (!locations->Find(out, id)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    Result RegisteredLocationResolverInterface::RefreshImpl(const ncm::ProgramId* excluding_ids, size_t num_ids) {
+        if (hos::GetVersion() < hos::Version_900) {
+            this->ClearRedirections();
+            return ResultSuccess();
+        }
+
+        if (num_ids == 0) {
+            this->ClearRedirections();
+        } else {
+            this->registered_program_locations.ClearExcluding(excluding_ids, num_ids);
```

Comment:
You have a helper that calls this on more than these two. Is that relevant here?